extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java

```diff
@@ -0,0 +1,130 @@
+        return AbstractCassandraClientProducer.class.getPackage().getName() + "."
+                + "CassandraClientProducer";
+    }
+
+    private void createCassandraClientProducerBean(
+            BuildProducer<GeneratedBeanBuildItem> generatedBean,
+            String cassandraClientProducerClassName) {
+
+        ClassOutput classOutput = new GeneratedBeanGizmoAdaptor(generatedBean);
+
+        try (ClassCreator classCreator = ClassCreator.builder().classOutput(classOutput)
+                .className(cassandraClientProducerClassName)
+                .superClass(AbstractCassandraClientProducer.class)
+                .build()) {
+            classCreator.addAnnotation(ApplicationScoped.class);
+
+            try (MethodCreator defaultCassandraClient = classCreator.getMethodCreator("createDefaultCassandraClient",
+                    CqlSession.class)) {
+                defaultCassandraClient.addAnnotation(ApplicationScoped.class);
+                defaultCassandraClient.addAnnotation(Produces.class);
+                defaultCassandraClient.addAnnotation(Default.class);
+
+                // make CqlSession as Unremovable bean
+                defaultCassandraClient.addAnnotation(Unremovable.class);
+
+                ResultHandle cassandraClientConfig = defaultCassandraClient.invokeVirtualMethod(
+                        MethodDescriptor.ofMethod(AbstractCassandraClientProducer.class, "getCassandraClientConfig",
+                                CassandraClientConfig.class),
+                        defaultCassandraClient.getThis());
+
+                defaultCassandraClient.returnValue(
+                        defaultCassandraClient.invokeVirtualMethod(
+                                MethodDescriptor.ofMethod(AbstractCassandraClientProducer.class, "createCassandraClient",
+                                        CqlSession.class,
+                                        CassandraClientConfig.class),
+                                defaultCassandraClient.getThis(),
+                                cassandraClientConfig));
+            }
+
+        }
+    }
+
+    @Record(RUNTIME_INIT)
+    @BuildStep
+    void configureRuntimePropertiesAndBuildClient(CassandraClientRecorder recorder,
+            CassandraClientConfig cassandraConfig,
+            ConfigurationBuildItem config) {
+        recorder.configureRuntimeProperties(cassandraConfig);
+    }
+
+    @BuildStep
+    @Record(value = RUNTIME_INIT, optional = true)
+    CassandraClientBuildItem cassandraClient(CassandraClientRecorder recorder) {
+        return new CassandraClientBuildItem(recorder.getClient());
+    }
+
+    @BuildStep
+    HealthBuildItem addHealthCheck(CassandraClientBuildTimeConfig buildTimeConfig) {
+        return new HealthBuildItem("io.quarkus.cassandra.runtime.health.CassandraHealthCheck",
+                buildTimeConfig.healthEnabled, "cassandra");
+    }
+
+    @BuildStep
+    RuntimeInitializedClassBuildItem runtimeMetadataManager() {
```

Comment:
`MetadataManager` must be initialized at runtime because it uses `Inet4Socket` address that cannot initialized at deployment time because of:
```
No instances of java.net.Inet4Address are allowed in the image heap as this class should be initialized at image runtime.
````